### DISCUSSION TITLE
Challenges and vue-dapp v1 Development Suggestions

### DISCUSSION THREAD

**Discussion by johnson86tw** [vu3th/vue-dapp#141] [General]
I'd like to initiate a discussion about the challenges that developers encounter while using vue-dapp to develop dapps. Currently, there is an opportunity to start preparing for the development of vue-dapp v1, and I hope to receive more development suggestions.

Here are some of my ideas:

- Adopting a monorepo approach to separate different functionalities, such as components and composables, to handle and publish them separately.
- Considering the creation of Nuxt modules as packages.
- Transforming external refs into Pinia plugins to import wallet data. (Using external refs might not be a standardized practice, even though it currently functions smoothly.)
- Abstracting the use of ethers from the project, allowing users to choose other web3 libraries like [web3.js](https://web3js.readthedocs.io/en/v1.10.0/) or [viem](https://viem.sh/).
- Considering using [@wagmi/core](https://github.com/wagmi-dev/wagmi/tree/main/packages/core) to replace the current self-made Connector in vue-dapp?

**Comment by re2005** [comment]
Nice initiative @chnejohnson I agree to every point you mentioned and would like to contribute as well.

Good luck!

**Comment by Zaskoda** [comment]
I very interested in this. I ended up not using vue-dapp in my project for reasons not worth detailing here. However, I did learn a lot from this project and applied it to the project I'm working on. I'm using Pinia to store the state of my contracts. It's a little complicated as my project is an EVM game with lots of different method calls across multiple contracts. So I end up with one store per contract plus another store to tie them altogether. Since other players can change the world, I basically re-build the game state with every new block. I do not know how to make this happen within Pinia alone. Right now I have a component that, with each new block, triggers an update in the Pinia store. I imagined other ways of doing it, but nothing felt like the "right" way. It seems like there should be a clean way to ensure that a Pinia store always has the latest contract state. I would love to see how you solve for this in vue-dapp.

**Comment by apuigsech** [comment]
I'm considering Nuxt as the framework of choice for the  [OpenLotto](https://github.com/apuigsech/openlotto-contracts) frontend. Your [vue-dapp](https://github.com/vu3th/vue-dapp) package, especially the [nuxt-dapp](https://github.com/vu3th/nuxt-dapp), seems to be a promising starting point.

Although I've encountered some challenges with nuxt-dapp, I'm not seeking solutions at this stage. Instead, I'm interested in learning about your developmental roadmap.

I've noticed that vue-dapp v1 is undergoing significant refactoring, and all ideas you listed look really good. Could you share your timeline for when you anticipate a stable release that we can use?

**Reply by johnson86tw** [reply]
@apuigsech Hi, I planned to finish the v1 upgrade by the end of 2023. Would that timeline work for your project? I can also try to accelerate the first stable release as MVP to meet your needs. I'm happy to see vue-dapp achieve broader adoption :)

**Reply by apuigsech** [reply]
I don't have a very tight timeline for the project, so the end of 2023 is fine. Although, obviously, the sooner the better :-)

We could explore a collaborative approach, using OpenLotto as a test case to uncover practical scenarios that could benefit both of us. But I have to tell that my expertise in Javascript/Typescript and frontend development is very limited, as my background is more focused on backend and systems programming (such as drivers and embedded systems), I'm certainly willing to contribute to the vue-dapp and nuxt-dapp projects where I can.

**Reply by johnson86tw** [reply]
Sounds great! We could explore collaboration opportunity. Take a look at OpenLotto; it looks interesting. I'm happy to help with that, and certainly, it could uncover practical scenarios that could improve the library.

There's a public vue-dapp Telegram group at https://t.me/+pLwZxOdgdBg3ZTRl, if you encounter any issues or problems during development, you can open an issue in the repo or send me messages in the group.
